File: apps/web/lib/api/radar.ts
import "server-only";

import { apiFetch } from "@/lib/server/api";

import type { RadarPage, RadarParams } from "./radar-types";

const SCALAR_KEYS = [
  "org_id",
  "score_min",
  "exchange",
  "anomaly_type",
  "regime",
  "volatility_min",
  "volatility_max",
  "q",
  "sort",
  "order",
  "limit",
  "cursor",
] as const satisfies readonly (keyof RadarParams)[];

const LIST_KEYS = ["status", "stage"] as const satisfies readonly (keyof RadarParams)[];

function radarQuery(params: RadarParams): string {
  const search = new URLSearchParams();
  for (const key of SCALAR_KEYS) {
    const value = params[key];
    if (value !== undefined) search.set(key, String(value));
  }
  for (const key of LIST_KEYS) {
    for (const value of params[key] ?? []) search.append(key, value);
  }
  const value = search.toString();
  return value ? `?${value}` : "";
}

/**
 * `GET /api/v1/radar` (`apps/api/hunter_api/routers/radar.py`, T2.6) -- one
 * row per scored `opportunities` episode (never a market row without one),
 * globally ranked. `org_id` only unlocks `in_position`/`risk_blocked` on
 * each row (`RadarPage.org_scoped`); every other field is identical for
 * every caller.
 */
export async function listRadar(params: RadarParams = {}): Promise<RadarPage> {
  return apiFetch<RadarPage>(`/api/v1/radar${radarQuery(params)}`);
}

File: apps/api/hunter_api/routers/radar.py
"""``GET /api/v1/radar`` — the cross-market opportunity radar (PIPELINE.md §5).

Not a tenant route: the scored opportunities are global (module docstring of
``repositories/radar.py``). ``org_id`` is an optional query parameter, not a
path segment, because the two fields it unlocks
(``in_position``/``risk_blocked``) are the only per-organization part of an
otherwise global response — see ``services/radar_org_derivation.py``.

The session is opened by ``routers/radar_common.py::analysis_scope`` in the
body rather than taken as a ``Depends``: see that module for why (one pooled
connection per request, never two).
"""

from __future__ import annotations

import uuid
from decimal import Decimal
from typing import TYPE_CHECKING, Annotated

from fastapi import APIRouter, Depends, Query, Request

from hunter_api.auth.rbac import CurrentPrincipal
from hunter_api.deps import get_redis
from hunter_api.repositories.base import MAX_PAGE_SIZE
from hunter_api.repositories.radar import RadarFilters
from hunter_api.routers.radar_common import analysis_scope
from hunter_api.schemas.radar import (
    MAX_SCORE,
    MAX_VOLATILITY,
    RadarPage,
    RadarSortKey,
    RadarStatusFilter,
    SortOrder,
)
from hunter_api.schemas.radar_coverage import RadarCoverageOut
from hunter_api.services.radar import build_radar_page, resolve_status_tokens
from hunter_api.services.radar_coverage import build_radar_coverage
from hunter_core.domain.enums import AnomalyType, MarketRegime, OpportunityStage

if TYPE_CHECKING:
    import redis.asyncio as redis_asyncio

router = APIRouter(prefix="/api/v1/radar", tags=["radar"])

Redis = Annotated["redis_asyncio.Redis", Depends(get_redis)]


@router.get("", response_model=RadarPage, summary="List the opportunity radar")
async def list_radar(
    request: Request,
    principal: CurrentPrincipal,
    org_id: uuid.UUID | None = None,
    score_min: Annotated[Decimal | None, Query(ge=0, le=MAX_SCORE)] = None,
    status: Annotated[list[RadarStatusFilter] | None, Query()] = None,
    stage: Annotated[list[OpportunityStage] | None, Query()] = None,
    exchange: Annotated[str | None, Query(max_length=32)] = None,
    anomaly_type: AnomalyType | None = None,
    regime: MarketRegime | None = None,
    volatility_min: Annotated[Decimal | None, Query(ge=0, le=MAX_VOLATILITY)] = None,
    volatility_max: Annotated[Decimal | None, Query(ge=0, le=MAX_VOLATILITY)] = None,
    q: Annotated[str | None, Query(max_length=64)] = None,
    sort: RadarSortKey = "score",
    order: SortOrder = "desc",
    limit: Annotated[int | None, Query(ge=1, le=MAX_PAGE_SIZE)] = None,
    cursor: str | None = None,
) -> RadarPage:
    async with analysis_scope(request, principal, org_id) as scope:
        statuses = resolve_status_tokens(status, has_org=scope.org_derivation is not None)
        filters = RadarFilters(
            score_min=score_min,
            statuses=statuses,
            stages=tuple(stage) if stage else (),
            exchange=exchange,
            anomaly_type=anomaly_type,
            regime=regime,
            volatility_min=volatility_min,
            volatility_max=volatility_max,
            symbol_query=q,
            sort=sort,
            order=order,
        )
        return await build_radar_page(
            scope.session,
            filters,
            limit=limit,
            cursor=cursor,
            org_derivation=scope.org_derivation,
        )


@router.get(
    "/coverage",
    response_model=RadarCoverageOut,
    summary="How much of the Radar exists yet (coverage, baselines, detectors)",
)
async def get_radar_coverage(
    request: Request, principal: CurrentPrincipal, redis: Redis
) -> RadarCoverageOut:
    """T3.46 (``.claude/state/notes-T3.46.md``): the Radar/Opportunities pages
    read this to say, in numbers, why the list has never had a candidate —
    global, not tenant-scoped (same reasoning as ``list_radar`` above), so no
    ``org_id`` is accepted here."""
    async with analysis_scope(request, principal) as scope:
        return await build_radar_coverage(scope.session, redis)
